lms/lti.py

```diff
@@ -0,0 +1,86 @@
+from dataclasses import dataclass
+from typing import Callable, Optional
+
+
+@dataclass
+class _ParamMapping:
+    key: str
+    """Name of the parameter in LTI1.3"""
+
+    sub: Optional[str] = None
+    """For nested values, name in the second object"""
+
+    function: Optional[Callable] = None
+    """If additional processing is needed, function to further process the value"""
+
+
+def _stringify_roles(lti_13_value):
+    """
+    Return roles as a comma separated list of values.
+
+    Roles are a list in LTI1.3 but comma separated in LTI1.1.
+    """
+    return ",".join(lti_13_value)
+
+
+class LTI13Params(dict):
+    """Dictionary subclass that translates LTI1.1 parameter names to the LTI1.3 ones."""
+
+    lti_param_mapping = {
```

Comment:
This is surprisingly small, not 100% of the parameters are here as I'm not covering the ones needed for configuring assignments but it's let's say 75% of the real size.

I think this approach has a few things going for it:

-  It's just python, all it's doing its translating one name to another so it doesn't need to be more complicated.
- It's all in one place and not scattered around different files.
- Works for marshmallow schema and more important, works with out them. 

Marshmallow schema expect a dictionary to `Schema().load` data through them so having this translation exposed as a dict works transparently as marshmallow is concern.

We have a problem where sometimes we need to access the LTI parameters without access to a schema. For example in pyramid predicates. Having this as a dict allows for that without any mayor app refactors.

Reply:
We've talked about this at great length, but long story short:

 * It's ideal if we don't mix the 1.1 and 1.3 things together in the same dict
 * There's a slightly simpler way to do this